alf/algorithms/sarsa_algorithm.py

```diff
@@ -0,0 +1,293 @@
+        if not isinstance(self._actor_network, DistributionNetwork):
+            action = tf.nest.map_structure(lambda a, ou: a + ou(),
+                                           policy_step.action,
+                                           self._ou_process)
+            policy_step = policy_step._replace(action=action)
+        return policy_step
+
+    def greedy_predict(self, time_step: ActionTimeStep, state=None, eps=0.1):
+        if isinstance(self._actor_network, DistributionNetwork):
+            return super().greedy_predict(time_step, state, eps=eps)
+        else:
+            return self._predict(time_step, state)
+
+    def rollout(self,
+                time_step: ActionTimeStep,
+                state: SarsaState,
+                is_train_step=False):
+        not_first_step = tf.not_equal(time_step.step_type, StepType.FIRST)
+        prev_critic, critic_state = self._critic_network(
+            inputs=(state.prev_observation, time_step.prev_action),
+            step_type=state.prev_step_type,
+            network_state=state.critic)
+        critic_state = tf.nest.map_structure(
+            lambda new_s, s: tf.where(not_first_step, new_s, s), critic_state,
+            state.critic)
+
+        action_distribution, action, actor_state = self._get_action(
+            time_step, self._actor_network, state.actor)
+
+        with tf.GradientTape(watch_accessed_variables=False) as tape:
+            tape.watch(action)
+            critic, _ = self._critic_network((time_step.observation, action),
+                                             step_type=time_step.step_type,
+                                             network_state=critic_state)
+        dqda = tape.gradient(critic, action)
+
+        def actor_loss_fn(dqda, action):
+            if self._dqda_clipping:
+                dqda = tf.clip_by_value(dqda, -self._dqda_clipping,
+                                        self._dqda_clipping)
+            loss = 0.5 * losses.element_wise_squared_loss(
+                tf.stop_gradient(dqda + action), action)
+            loss = tf.reduce_sum(loss, axis=list(range(1, len(loss.shape))))
+            return loss
+
+        actor_loss = tf.nest.map_structure(actor_loss_fn, dqda, action)
+        actor_loss = tf.add_n(tf.nest.flatten(actor_loss))
+
+        _, target_action, target_actor_state = self._get_action(
+            time_step, self._target_actor_network, state.target_actor)
+
+        target_critic, target_critic_state = self._target_critic_network(
+            (time_step.observation, target_action),
+            step_type=time_step.step_type,
+            network_state=state.target_critic)
+
+        prev_return = tf.stop_gradient(time_step.reward +
+                                       time_step.discount * target_critic)
+        info = SarsaInfo(
+            actor_loss=actor_loss, critic=prev_critic, returns=prev_return)
+
+        if not is_train_step and not isinstance(self._actor_network,
```

Comment:
use `not isinstance(action_distribution, tf.Tensor)` like `common.to_distribution` seems better

Reply:
It's more trouble because action_distribution might be a nest.

Reply:
i see